File: Src/ECScape.Core/Engine/GameTimer.cs
using System.Diagnostics;

namespace ECScape.Core.Engine;

internal sealed class GameTimer
{
    private readonly Stopwatch stopwatch = Stopwatch.StartNew();
    private long lastFrameTimeMs;
    private const double TargetFrameTimeMs = 1000.0 / Configuration.TargetFrameRate;

    public float DeltaTime { get; private set; }

    public GameTimer()
    {
        lastFrameTimeMs = stopwatch.ElapsedMilliseconds;
    }

    public void Update()
    {
        var currentTimeMs = stopwatch.ElapsedMilliseconds;
        DeltaTime = (currentTimeMs - lastFrameTimeMs) / 1000.0f;
        lastFrameTimeMs = currentTimeMs;
    }

    public async Task LimitFrameRate()
    {
        var frameEndTimeMs = stopwatch.ElapsedMilliseconds;
        var frameProcessingTimeMs = frameEndTimeMs - lastFrameTimeMs;
        var sleepTimeMs = TargetFrameTimeMs - frameProcessingTimeMs;

        if (sleepTimeMs > 1)
        {
            await Task.Delay((int)sleepTimeMs);
        }
    }
}

File: src/core/GameTimer.js
import Configuration from './configuration.js';

class GameTimer {
    constructor() {
        this.lastFrameTimeMs = performance.now();
        this.targetFrameTimeMs = 1000 / Configuration.TargetFrameRate;
        this.deltaTime = 0;
    }

    update() {
        const currentTimeMs = performance.now();
        this.deltaTime = (currentTimeMs - this.lastFrameTimeMs) / 1000;
        this.lastFrameTimeMs = currentTimeMs;
    }

    async limitFrameRate() {
        const frameEndTimeMs = performance.now();
        const frameProcessingTimeMs = frameEndTimeMs - this.lastFrameTimeMs;
        const sleepTimeMs = this.targetFrameTimeMs - frameProcessingTimeMs;

        if (sleepTimeMs > 1) {
            await new Promise(resolve => setTimeout(resolve, sleepTimeMs));
        }
    }
}

export default GameTimer;